streams/client/client.go

```diff
@@ -178,82 +208,132 @@ func (c *Client) send(cmdName string, body command.Body) (peerCore.ID, cr.Respon
-			}
+func (c *Client) connect(peer peerCore.AddrInfo) (network.Stream, bool, error) {
+	switch c.node.Peer().Network().Connectedness(peer.ID) {
+	case network.Connected:
+	case network.CanConnect, network.NotConnected:
+		go c.node.Peer().Connect(c.ctx, peer)
+		return nil, true, nil
+	default:
+		return nil, false, nil
+	}
 
-			var err error
-			strm, err = c.node.Peer().NewStream(network.WithNoDial(c.ctx, "application ensured connection to peer exists"), pid, proto)
-			if err != nil {
-				logger.Errorf("starting stream to `%s`;`%s` failed with: %s", pid.String(), c.path, err)
-				continue
-			}
-		case <-ctx.Done():
-			return "", nil, fmt.Errorf("%s: finding peers for command `%s`.`%s` timed out!: %s", c.node.ID(), c.path, cmdName, ctx.Err())
-		}
+	strm, err := c.node.Peer().NewStream(network.WithNoDial(c.ctx, "application ensured connection to peer exists"), peer.ID, protocol.ID(c.path))
+	if err != nil {
+		logger.Errorf("starting stream to `%s`;`%s` failed with: %s", peer.ID.String(), c.path, err)
+		return nil, false, err
 	}
 
-	defer strm.Close()
+	return strm, false, nil
+}
+
+func (c *Client) sendTo(strm stream, deadline time.Time, cmdName string, body command.Body) response {
+	cmd := command.New(cmdName, body)
 
-	if err := strm.SetWriteDeadline(min(time.Now().Add(SendTimeout), strmDD)); err != nil {
-		return "", nil, fmt.Errorf("setting write deadline at `%s`;`%s` failed with: %s", pid.String(), c.path, err)
+	if err := strm.SetWriteDeadline(min(time.Now().Add(SendTimeout), deadline)); err != nil {
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("setting write deadline failed with: %s", err),
+		}
 	}
 
 	if err := cmd.Encode(strm); err != nil {
-		dl, ok := ctx.Deadline()
-		return "", nil, fmt.Errorf("encoding at `%s`;`%s` t(%s::%v) failed with: %s", pid.String(), c.path, dl.String(), ok, err)
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("seding command `%s(%s)` failed with: %s", cmd.Command, c.path, err),
+		}
 	}
 
-	if err := strm.SetReadDeadline(min(time.Now().Add(RecvTimeout), strmDD)); err != nil {
-		return "", nil, fmt.Errorf("setting read deadline at `%s`;`%s` failed with: %s", pid.String(), c.path, err)
+	if err := strm.SetReadDeadline(min(time.Now().Add(RecvTimeout), deadline)); err != nil {
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("setting read deadline failed with: %s", err),
+		}
 	}
 
-	response, err := cr.Decode(strm)
+	resp, err := cr.Decode(strm)
 	if err != nil {
-		dl, ok := ctx.Deadline()
-		return "", nil, fmt.Errorf("decoding at `%s`;`%s` t(%s::%v) failed with: %s", pid.String(), c.path, dl.String(), ok, err)
+		return response{
+			ID:    strm.ID,
+			error: fmt.Errorf("recv response of `%s(%s)` failed with: %s", cmd.Command, c.path, err),
+		}
 	}
 
-	if v, k := response["error"]; k {
-		return pid, cr.Response{}, errors.New(fmt.Sprint(v))
+	if v, k := resp["error"]; k {
+		return response{
+			ID:    strm.ID,
+			error: errors.New(fmt.Sprint(v)),
```

Comment:
Is this ok? Shouldnt it be safer to fmt.Sprintf("%v",v)

Reply:
Nope.